concordium-consensus/src/Concordium/KonsensusV1/TreeState/StartUp.hs

```diff
@@ -209,3 +231,88 @@ loadSkovData _runtimeParameters = do
                 }
     let _statistics = Stats.initialConsensusStatistics
     return SkovData{..}
+
+-- |Load the certified blocks from the low-level database into the tree state.
+-- This caches their block states, adds them to the block table and branches,
+-- adds their transactions to the transaction table and pending transaction table,
+-- updates the highest certified block, and records block signature witnesses and
+-- checked quorum certificates for the blocks.
+--
+-- This also sets the previous round timeout if the low level state records that it timed out.
+--
+-- This should be called on the result of 'loadSkovData' after the transaction table has
+-- been initialised for the last finalized block.
+loadCertifiedBlocks ::
+    forall m.
+    ( MonadThrow m,
+      LowLevel.MonadTreeStateStore m,
+      MonadIO m,
+      BlockStateStorage m,
+      GSTypes.BlockState m ~ PBS.HashedPersistentBlockState (MPV m),
+      MonadState (SkovData (MPV m)) m,
+      TimeMonad m
+    ) =>
+    m ()
+loadCertifiedBlocks = do
+    certBlocks <- LowLevel.lookupCertifiedBlocks
+    mapM_ loadCertBlock certBlocks
+    oLastTimeout <- use $ persistentRoundStatus . prsLatestTimeout
+    forM_ oLastTimeout $ \lastTimeout -> do
+        curRound <- use $ roundStatus . rsCurrentRound
+        when (tcRound lastTimeout >= curRound) $ do
+            highCB <- use $ roundStatus . rsHighestCertifiedBlock
+            unless
+                ( cbRound highCB < tcRound lastTimeout
+                    && cbRound highCB >= tcMaxRound lastTimeout
+                    && cbEpoch highCB >= tcMaxEpoch lastTimeout
+                    && cbEpoch highCB <= 2 + tcMinEpoch lastTimeout
+                )
+                $ throwM . TreeStateInvariantViolation
+                $ "Missing certified block consistent with last timeout certificate"
+            roundStatus . rsPreviousRoundTimeout
+                .= Present
+                    RoundTimeout
+                        { rtTimeoutCertificate = lastTimeout,
+                          rtCertifiedBlock = highCB
+                        }
+            roundStatus . rsCurrentRound .= tcRound lastTimeout + 1
+    rs <- use roundStatus
+    let expectedCurrentRound
+            | Present prevTO <- rs ^. rsPreviousRoundTimeout = 1 + tcRound (rtTimeoutCertificate prevTO)
+            | otherwise = 1 + cbRound (rs ^. rsHighestCertifiedBlock)
+    unless (expectedCurrentRound == rs ^. rsCurrentRound) $
+        throwM . TreeStateInvariantViolation $
+            "The current round does not match the expected round."
```

Comment:
I don't think this check makes sense in the presence of rollbacks.

Suppose we have no timeouts, and for some reason or database (block state) gets corrupt. Then we'll have deleted a number of certified blocks first in `initialiseExistingSkov`.

But if I understand correctly how `rsCurrentRound` is defined, it is always at least the one stored in the `roundStatusStore`. So these two together cannot really work unless I'm misunderstanding something.

I suppose we don't want to roll back the persistent round status, so we don't send contradicting messages.

Reply:
I have now changed how `rsCurrentRound` is defined, so it is the highest round for which we have a valid QC or TC for the previous round (i.e. it is justified in the current state). Additional checks should be used to prevent baking and/or signing for "old" rounds.